### DISCUSSION TITLE
Not portable types

### DISCUSSION THREAD

**Discussion by sohrab-** [amannn/next-intl#1014] [Q&A] [answered]
When using various functions, we instantiate variables that do not seem to have a portable inferred type.

For example:

```tsx
export const { Link, redirect, usePathname, useRouter } =
  createSharedPathnamesNavigation({
    locales,
    localePrefix,
  });
```

results in 

```
The inferred type of 'Link' cannot be named without a reference to '../node_modules/next-intl/dist/types/src/shared/types'. This is likely not portable. A type annotation is necessary.ts(2742)
The inferred type of 'useRouter' cannot be named without a reference to '../node_modules/next-intl/dist/types/src/navigation/react-client/useBaseRouter'. This is likely not portable. A type annotation is necessary.ts(2742)
```

Similarly,

```tsx
const requestConfig = getRequestConfig(
  async ({ locale }) => { ...  }
);
```

results in 

```
The inferred type of 'requestConfig' cannot be named without a reference to '../../node_modules/.pnpm/use-intl@3.11.1_react@18.2.0/node_modules/use-intl/dist/types/src/core/TimeZone'. This is likely not portable. A type annotation is necessary.ts(2742)
The inferred type of 'requestConfig' cannot be named without a reference to './node_modules/next-intl/dist/types/src/server/react-server/getRequestConfig'. This is likely not portable. A type annotation is necessary.ts(2742)
```

For the latter, I seem to be able to get rid of it by typing it as `ReturnType<typeof getRequestConfig>` but the former is not as easily resolved.

Do the `next-intl` packages need to export these additional types, e.g. `AllLocales`, `TimeZone`, etc.? Or am I doing something wrong here?

**Comment by amannn** [comment]
Can you share your `tsconfig.json` or ideally a full reproduction in [a bug report](https://github.com/amannn/next-intl/issues/new?assignees=&labels=bug%2Cunconfirmed&projects=&template=01_bug_report.yml)?

**Reply by thebossjon** [reply]
[TypeScript/issues/47663](https://github.com/microsoft/TypeScript/issues/47663)

**Reply by sohrab-** [reply]
Here are my ts configs. I am working on a brand new Next.js app with just [standard setup](https://next-intl-docs.vercel.app/docs/getting-started/app-router) of next-intl.

```jsonc
// tsconfig.json
{
  "extends": "@repo/typescript-config/nextjs.json",
  "compilerOptions": {
    "plugins": [
      {
        "name": "next"
      }
    ]
  },
  "include": [
    "next-env.d.ts",
    "next.config.js",
    "**/*.js",
    "**/*.ts",
    "**/*.tsx",
    ".next/types/**/*.ts"
  ],
  "exclude": ["node_modules"]
}

// @repo/typescript-config/nextjs.json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Next.js",
  "extends": "./base.json",
  "compilerOptions": {
    "plugins": [{ "name": "next" }],
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "allowJs": true,
    "jsx": "preserve",
    "noEmit": true,
    "paths": {
      "@ui/*": ["../../packages/ui/src/*"]
    }
  }
}

// @repo/typescript-config/base.json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Default",
  "compilerOptions": {
    "declaration": true,
    "declarationMap": true,
    "esModuleInterop": true,
    "incremental": false,
    "isolatedModules": true,
    "lib": ["es2022", "DOM", "DOM.Iterable"],
    "module": "NodeNext",
    "moduleDetection": "force",
    "moduleResolution": "NodeNext",
    "noUncheckedIndexedAccess": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "strict": true,
    "target": "ES2022"
  }
}

```


I'll submit a bug report the first chance I get.

**Reply by amannn** [reply, marked as answer]
I've isolated the issue to `"declaration": true`. If you remove that, the error should be gone. Note that this option isn't set by default in a new Next.js app.

If there's anything that `next-intl` can do to be more compatible with this config option I'd generally be open to discuss, but as far as I understand, this setting is rather intended to be used in libraries rather than apps.

Hope this helps!

**Reply by sohrab-** [reply]
Thanks for that.

Yeah, it is odd coz they do seem to set it in [turborepo starter](https://github.com/vercel/turbo/blob/04248f45b84789488324c1d9753c3b4e49213529/examples/basic/packages/typescript-config/base.json#L5).

I haven't personally groked [this](https://github.com/microsoft/TypeScript/pull/58176#issuecomment-2052698294) to figure out what changes would be required to make sure this works while leaving the `declaration` option on.

**Reply by MrMilan** [reply]
I recently migrated my Next.js project to Turborepo and encountered the same issue. I wasn't sure what was causing it and whether it was a valid problem.
Thanks a lot for explanation and solution.

**Reply by vaishnavi-jadhav-amla** [reply]
@sohrab- iam facing the same issue and Iam also using turorepo. Were you able to find any solution. I encountered this issue while using getRequestConfig option. Although it worked by removing the declaration option. But wanted to understand, is this the right way

**Reply by ol-ts** [reply]
What if you can't set `declaration` to `false`? 
We reference some packages in our tsconfig and use
``` "compilerOptions": {"composite": true}``` which sets the `declaration` to true. So we can't have both `composite: true` and `declaration: false` as these two options are incompatible.

**Reply by osaton** [reply]
This is a problem when developing a library for Next.js that uses `next-intl` internally, as the types are not exportable.

**Reply by sohrab-** [reply]
Turning off declarations is more of a workaround. I think using next-intl in an internal library and re-exporting parts of it is a pretty valid use case so it should be supported. If someone has a reproduction repo handy, can you please [submit a bug](https://github.com/amannn/next-intl/issues/new?assignees=&labels=bug%2Cunconfirmed&projects=&template=01_bug_report.yml).

**Reply by amannn** [reply]
>  I think using next-intl in an internal library and re-exporting parts of it is a pretty valid use case so it should be supported.

Yep, I think that can be agreed to.

If somebody has experience with resolving this issue, I'd also be happy to review a PR!

**Reply by amannn** [reply]
@osaton fixed this in https://github.com/amannn/next-intl/pull/1509 👏
